### PR TITLE
feat: overview async

### PR DESCRIPTION
**Pull request by thealligatorking**
# Overview Async

* Coroutines for `_get_neurons_for_netuid` & `_get_de_registered_stake_for_coldkey_wallet`
image

### CODE REVIEW

File: bittensor/commands/overview.py [hunk 1]
```diff
@@ -612,81 +607,56 @@ def overview_sort_function(row):
...
-            all_staked_hotkeys = filter(_filter_stake_info, all_stake_info_for_coldkey)
-            result = [
-                (
-                    stake_info.hotkey_ss58,
-                    stake_info.stake.tao,
-                )  # stake is a Balance object
-                for stake_info in all_staked_hotkeys
-            ]
+        # Pull all stake for our coldkey
+        all_stake_info_for_coldkey = subtensor.get_stake_info_for_coldkey(
```
Comment (basfroman) [inline bittensor/commands/overview.py]: should be awaited

File: bittensor/commands/overview.py [hunk 2]
```diff
@@ -612,81 +607,56 @@ def overview_sort_function(row):
...
-                subtensor.close()
-                bittensor.logging.debug("closing subtensor connection")
+        ## Filter out hotkeys that are in our wallets
+        ## Filter out hotkeys that are delegates.
+        def _filter_stake_info(stake_info: "bittensor.StakeInfo") -> bool:
+            if stake_info.stake == 0:
+                return False  # Skip hotkeys that we have no stake with.
+            if stake_info.hotkey_ss58 in all_hotkey_addresses:
+                return False  # Skip hotkeys that are in our wallets.
+            return not subtensor.is_hotkey_delegate(hotkey_ss58=stake_info.hotkey_ss58)
```
Comment (basfroman) [inline bittensor/commands/overview.py]: should be awaited

File: bittensor/commands/overview.py [hunk 3]
```diff
@@ -612,81 +607,56 @@ def overview_sort_function(row):
...
 
-        except Exception as e:
-            return coldkey_wallet, [], "Error: {}".format(e)
-        finally:
-            if "subtensor" in locals():
-                subtensor.close()
-                bittensor.logging.debug("closing subtensor connection")
+        ## Filter out hotkeys that are in our wallets
+        ## Filter out hotkeys that are delegates.
+        def _filter_stake_info(stake_info: "bittensor.StakeInfo") -> bool:
```
Comment (basfroman) [inline bittensor/commands/overview.py]: should be async

File: bittensor/commands/overview.py [hunk 4]
```diff
@@ -612,81 +607,56 @@ def overview_sort_function(row):
...
-                if uid is not None:
-                    nn = all_neurons[uid]
-                    result.append(nn)
-        except Exception as e:
-            return netuid, [], "Error: {}".format(e)
-        finally:
-            if "subtensor" in locals():
-                subtensor.close()
-                bittensor.logging.debug("closing subtensor connection")
+        all_neurons: List["bittensor.NeuronInfoLite"] = subtensor.neurons_lite(
```
Comment (basfroman) [inline bittensor/commands/overview.py]: should be awaited

File: bittensor/commands/overview.py [hunk 5]
```diff
@@ -258,9 +255,9 @@ def _run(cli: "bittensor.cli", subtensor: "bittensor.subtensor"):
...
                         continue
-                    total_coldkey_stake_from_metagraph[neuron.coldkey] += (
-                        neuron.stake_dict[neuron.coldkey]
-                    )
+                    total_coldkey_stake_from_metagraph[
+                        neuron.coldkey
+                    ] += neuron.stake_dict[neuron.coldkey]
                     checked_hotkeys.add(neuron.hotkey)
 
             alerts_table = Table(show_header=True, header_style="bold magenta")
```
Comment (RomanCh-OT) [inline bittensor/commands/overview.py:263]: subtensor.get_total_stake_for_coldkey below should be awaited. total_coldkey_stake_from_chain is coroutine

**Review by basfroman — COMMENTED** [review]
Since github doesn't create the commends on the unmodified lines, I put it here.
The following list of methods and properties have to be awaited across the code:
- subtensor.get_total_stake_for_coldkey()
- subtensor.get_balance()
- subtensor.block
- subtensor.get_all_subnet_netuids()
- subtensor.tempo()
- subtensor.neurons_lite()
- subtensor.get_stake_info_for_coldkey()
- subtensor.is_hotkey_delegate()

After all changes the inner function `_filter_stake_info` become to async too and should be awaited.

**Comment by basfroman** [comment]
also, pls check `ruff` formatting

File: bittensor/commands/overview.py [hunk 6]
```diff
@@ -203,9 +203,9 @@ def _run(cli: "bittensor.cli", subtensor: "bittensor.subtensor"):
 
         # Pull neuron info for all keys.
         neurons: Dict[str, List[bittensor.NeuronInfoLite]] = {}
-        block = subtensor.block
+        block = await subtensor.block
```
Comment (basfroman) [inline bittensor/commands/overview.py]: `subtensor.block` and `subtensor.get_all_subnet_netuids()` could be gathered

File: bittensor/commands/overview.py [hunk 7]
```diff
@@ -103,7 +103,7 @@ def _get_total_balance(
                     cold_wallet.coldkeypub_file.exists_on_device()
                     and not cold_wallet.coldkeypub_file.is_encrypted()
                 ):
-                    total_balance = total_balance + subtensor.get_balance(
+                    total_balance = total_balance + await subtensor.get_balance(
```
Comment (basfroman) [inline bittensor/commands/overview.py:113]: line#117 has one more `subtensor.get_balance` call which have to be awaited

**Comment by thewhaleking** [comment]
Overview Command

I think this is a better approach.

**Comment by basfroman** [comment]
> Overview Command
> 
> I think this is a better approach.

will you able to use that approach with tui? I mean won't this block tui?

**Comment by thewhaleking** [comment]
> will you able to use that approach with tui? I mean won't this block tui?

Why would it block Commander?

**Comment by basfroman** [comment]
> > will you able to use that approach with tui? I mean won't this block tui?
> 
> Why would it block Commander?

I'm not saying this, just asking. I like the proposed option, but due to the fact that it is a big restructuring of logic, I'm just care about it.

**Comment by thewhaleking** [comment]
> I'm not saying this, just asking. I like the proposed option, but due to the fact that it is a big restructuring of logic, I'm just care about it.

It should not block Commander. The logic needs to change in a pretty substantial way anyway, so I think this is the optimal path.

File: bittensor/subtensor.py [hunk 8]
```diff
@@ -4918,7 +4956,7 @@ async def neurons_lite(
         This function offers a quick overview of the neuron population within a subnet, facilitating
         efficient analysis of the network's decentralized structure and neuron dynamics.
         """
-        hex_bytes_result = await self.query_runtime_api(
+        hex_bytes_result = self.query_runtime_api(
```
Comment (basfroman) [inline bittensor/subtensor.py]: `self.query_runtime_api` should be awaited

File: bittensor/commands/overview.py [hunk 9]
```diff
@@ -179,7 +195,7 @@ def _process_neuron_results(
                 neurons[str(netuid)] = neurons_result
         return neurons
 
-    def _run(cli: "bittensor.cli", subtensor: "bittensor.subtensor"):
+    async def _run(cli: "bittensor.cli", subtensor: "bittensor.subtensor"):
```
Comment (RomanCh-OT) [inline bittensor/commands/overview.py:198]: I would suggest replace `cli` to `self` (PEP8)

File: bittensor/commands/overview.py [hunk 10]
```diff
@@ -203,9 +219,9 @@ def _run(cli: "bittensor.cli", subtensor: "bittensor.subtensor"):
 
         # Pull neuron info for all keys.
         neurons: Dict[str, List[bittensor.NeuronInfoLite]] = {}
-        block = subtensor.block
-
-        netuids = subtensor.get_all_subnet_netuids()
+        block, netuids = await asyncio.gather(
+            subtensor.block(), subtensor.get_all_subnet_netuids()
```
Comment (basfroman) [inline bittensor/commands/overview.py]: `subtensor.block` if property (call without `()`)

File: bittensor/commands/overview.py [hunk 11]
```diff
@@ -232,45 +248,34 @@ def _run(cli: "bittensor.cli", subtensor: "bittensor.subtensor"):
...
-                    OverviewCommand._get_neurons_for_netuid,
-                    [(copy_config, netuid, all_hotkey_addresses) for netuid in netuids],
-                )
-                executor.shutdown(wait=True)  # wait for all complete
+            # Pull neuron info for all keys.:
 
-                neurons = OverviewCommand._process_neuron_results(
-                    results, neurons, netuids
-                )
+            results = OverviewCommand._get_neurons_for_netuids(
```
Comment (basfroman) [inline bittensor/commands/overview.py]: `_get_neurons_for_netuids` is awaitable object

File: bittensor/commands/overview.py [hunk 12]
```diff
@@ -612,81 +615,86 @@ def overview_sort_function(row):
         console.print(grid, width=cli.config.get("width", None))
 
     @staticmethod
-    def _get_neurons_for_netuid(
-        args_tuple: Tuple["bittensor.Config", int, List[str]],
-    ) -> Tuple[int, List["bittensor.NeuronInfoLite"], Optional[str]]:
-        subtensor_config, netuid, hot_wallets = args_tuple
-
+    def _map_hotkey_to_neurons(
```
Comment (basfroman) [inline bittensor/commands/overview.py:626]: pls add return value type annotation

File: bittensor/commands/overview.py [hunk 13]
```diff
@@ -612,81 +615,86 @@ def overview_sort_function(row):
...
         return netuid, result, None
 
     @staticmethod
-    def _get_de_registered_stake_for_coldkey_wallet(
-        args_tuple,
+    async def _fetch_neuron_for_netuid(
+        netuid: int, subtensor: "bittensor.subtensor"
+    ) -> Tuple[int, "bittensor.NeuronInfoLite"]:
+        neurons = await subtensor.neurons_lite_for_uid(uid=netuid)
+        return (netuid, neurons)
```
Comment (basfroman) [inline bittensor/commands/overview.py]: you can remove parentheses, just `return netuid, neurons`

File: bittensor/commands/overview.py [hunk 14]
```diff
@@ -612,81 +615,86 @@ def overview_sort_function(row):
...
+            return netuid, [], f"Error: {e}"
 
         return netuid, result, None
 
     @staticmethod
-    def _get_de_registered_stake_for_coldkey_wallet(
-        args_tuple,
+    async def _fetch_neuron_for_netuid(
+        netuid: int, subtensor: "bittensor.subtensor"
+    ) -> Tuple[int, "bittensor.NeuronInfoLite"]:
```
Comment (basfroman) [inline bittensor/commands/overview.py]: function return annotation is `Tuple[int,  list["bittensor.NeuronInfoLite"]]`

File: bittensor/commands/overview.py [hunk 15]
```diff
@@ -612,81 +615,86 @@ def overview_sort_function(row):
...
+        neurons = await subtensor.neurons_lite_for_uid(uid=netuid)
+        return (netuid, neurons)
+
+    @staticmethod
+    async def _fetch_all_neurons(
+        netuids: List[int], subtensor
+    ) -> List[Tuple[int, List["bittensor.NeuronInfoLite"]]]:
+        return await asyncio.gather(
```
Comment (basfroman) [inline bittensor/commands/overview.py]: wrap return into the list (as says in function annotation)
```python
        return list(await asyncio.gather(
            *[
                OverviewCommand._fetch_neuron_for_netuid(netuid, subtensor)
                for netuid in netuids
            ]
        ))
```

File: bittensor/commands/overview.py [hunk 16]
```diff
@@ -612,81 +615,86 @@ def overview_sort_function(row):
...
+            *[
+                OverviewCommand._fetch_neuron_for_netuid(netuid, subtensor)
+                for netuid in netuids
+            ]
+        )
+
+    @staticmethod
+    async def _get_neurons_for_netuids(
+        subtensor: "bittensor.subtensor", netuids: List[int], hot_wallets: List[str]
+    ) -> Tuple[int, List["bittensor.NeuronInfoLite"], Optional[str]]:
```
Comment (basfroman) [inline bittensor/commands/overview.py]: nitpick: a function's return value annotation does not match its return type

File: bittensor/subtensor.py [hunk 17]
```diff
@@ -4900,6 +4901,43 @@ async def neuron_for_uid_lite(
 
         return NeuronInfoLite.from_vec_u8(bytes_result)  # type: ignore
 
+    def decode_scale_bytes(self, return_type, scale_bytes, custom_rpc_type_registry):
```
Comment (basfroman) [inline bittensor/subtensor.py]: let's move `decode_scale_bytes` function to `bittensor/commands/utils.py` or `bittensor/commands/overview.py` as a helper function

**Review by basfroman — CHANGES_REQUESTED** [review]

File: bittensor/commands/overview.py [hunk 18]
```diff
@@ -179,7 +202,7 @@ def _process_neuron_results(
                 neurons[str(netuid)] = neurons_result
         return neurons
```
Comment (thewhaleking) [inline bittensor/commands/overview.py:204]: `@staticmethod`

**Review by thewhaleking — APPROVED** [review]
Looks good man! Just make that one method a staticmethod, and you're good to go!

**Pull request merged**